File: 백준/Gold/2179. 비슷한 단어/비슷한 단어.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine().trim());

        List<Word> list = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            list.add(new Word(br.readLine().trim(), i));
        }

        Collections.sort(list);

        String prefix = "";
        int minIdx = Integer.MAX_VALUE;

        for (int i = 0; i < n - 1; i++) {
            Word w1 = list.get(i);
            Word w2 = list.get(i + 1);

            int len = getPrefixLen(w1.text, w2.text);

            if (len > prefix.length() || (len == prefix.length() && Math.min(w1.idx, w2.idx) < minIdx)) {
                prefix = w1.text.substring(0, len);
                minIdx = Math.min(w1.idx, w2.idx);
            }
        }

        List<Word> res = new ArrayList<>();
        for (Word w : list) {
            if (w.text.startsWith(prefix)) {
                res.add(w);
            }
        }

        res.sort(Comparator.comparingInt(w -> w.idx));

        System.out.println(res.get(0).text);
        System.out.println(res.get(1).text);
    }

    private static int getPrefixLen(String s1, String s2) {
        int len = Math.min(s1.length(), s2.length());
        int i = 0;
        while (i < len && s1.charAt(i) == s2.charAt(i)) {
            i++;
        }
        return i;
    }

    // 단어 + 인덱스 저장
    static class Word implements Comparable<Word> {
        String text;
        int idx;

        public Word(String text, int idx) {
            this.text = text;
            this.idx = idx;
        }

        @Override
        public int compareTo(Word o) {
            return this.text.compareTo(o.text);
        }
    }
}

File: 백준/Gold/2179. 비슷한 단어/비슷한 단어.py
import sys
input = sys.stdin.readline

n = int(input())
words = [(input().strip(), i) for i in range(n)]
res = ''
answer = float('inf')

words.sort()
for i in range(n - 1):
    if words[i][0] == words[i + 1][0]:  #중복 단어 제거
        continue

    min_len = min(len(words[i][0]), len(words[i + 1][0]))
    min_idx = min(words[i][1], words[i + 1][1])

    j = 0
    while j < min_len and words[i][0][j] == words[i + 1][0][j]:
        j += 1

    if len(res) < j or (len(res) == j and min_idx < answer):
        res = words[i][0][:j]
        answer = min_idx 

answer_list = []
if res:
    for word, idx in words:
        if word[:len(res)] == res:
            answer_list.append((word, idx))
answer_list.sort(key=lambda x: x[1])
for word, _ in answer_list[:2]:
    print(word)
